lib/group_lookup.rb

```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+class GroupLookup
+  def initialize(group_ids = [])
+    @group_ids = group_ids.tap(&:flatten!).tap(&:compact!).tap(&:uniq!)
```

Comment:
I wonder, how is this different than this? 🤔 

```suggestion
    @group_ids = group_ids.flatten.compact.uniq
```

The ! methods change the object "inline" so they don't create any copy but I wonder if there are any performance differences here? I like the one above which is much easier to read/understand. Though if there are performances improvements to using the tag + ! combo, then it's fine to go with it.

Reply:
Initially, I copied that code from [`user_lookup.rb`](https://github.com/discourse/discourse/commit/fab8b8649e6994d7e749c83247b4ffc0f6d3f0ec#diff-c376db47a980f33fbf31e5278dc0ace95b993791063460ec9e762c394dd8819eR6) to group_lookup file :)

In a GitHub code search, I can see many other coders are using the tap + ! combo. So it must have a performance improvement. But I'm unable to find articles about it. @davidtaylorhq can you please explain this combo that you introduced in [user_lookup](https://github.com/discourse/discourse/commit/fab8b8649e6994d7e749c83247b4ffc0f6d3f0ec#diff-c376db47a980f33fbf31e5278dc0ace95b993791063460ec9e762c394dd8819eR6) file?

Reply:
I agree. We don't need this optimization in this case since the number of groups is way less. I will commit the above code suggestion.

Anyway, I still want to hear about the reason :)